build failure with libcxx 19.1.7 on OpenBSD

### What is the bug?

OpenBSD recently moved from building with llvm 16 to llvm 19, and is in the process of updating libcxx too. @botovq is doing a test bulk (i'm not taking credit for anything, just being the messenger), and found that gdal 3.11 fails to build with the following error:
```
/usr/include/c++/v1/__locale_dir/locale_base_api/bsd_locale_fallbacks.h:109:15: error: use of undeclared identifier 'vasprintf'; did you mean 'vsprintf'?
```
apparently libcxx 19.1.7 pulls in vasprintf, for which we need _XOPEN_SOURCE bumped to 800 ?

inlined patch below apparently helps:
```
Index: port/cpl_conv.cpp
--- port/cpl_conv.cpp.orig
+++ port/cpl_conv.cpp
@@ -14,7 +14,7 @@
 #include "cpl_config.h"
 
 #ifdef HAVE_USELOCALE
-// For uselocale, define _XOPEN_SOURCE = 700
// For uselocale, define _XOPEN_SOURCE = 800
 // but on Solaris, we don't have uselocale and we cannot have
 // std=c++11 with _XOPEN_SOURCE != 600
 #if defined(__sun__) && __cplusplus >= 201103L
@@ -28,7 +28,7 @@
 #ifdef _XOPEN_SOURCE
 #undef _XOPEN_SOURCE
 #endif
-#define _XOPEN_SOURCE 700
+#define _XOPEN_SOURCE 800
 #endif
 #endif 
```

### Steps to reproduce the issue

build on OpenBSD with libcxx 19.1.7

### Versions and provenance

OpenBSD-current, gdal 3.11

### Additional context

_No response_

Specifically, it comes from

https://github.com/llvm/llvm-project/blob/cd708029e0b2869e80abe31ddb175f7c35361f90/libcxx/include/__locale_dir/locale_base_api/bsd_locale_fallbacks.h#L109

Here's the detailed error:

```
FAILED: port/CMakeFiles/cpl.dir/cpl_conv.cpp.o 
/tmp/pobj/gdal-3.11.0/bin/c++ -DDONT_DEPRECATE_SPRINTF -DENABLE_DEFLATE64 -DGDAL_COMPILATION -DHAVE_CURL -DHAVE_ICONV -DHAVE_LIBARCHIVE -DHAVE_LIBDEFLATE -DHAVE_LIBXML2 -DHAVE_LIBZ -DHAVE_LZ4 -DHAVE_LZMA -DHAVE_OPENSSL_CRYPTO -DHAVE_PREAD_BSD -DHAVE_ZLIB -DHAVE_ZLIB_H -DHAVE_ZSTD -DINST_DATA=\"/usr/local/share/gdal\" -DSIZEOF_OFF_T=8 -DSYSCONFDIR=\"/usr/local/etc\" -DZIP_SUPPORT -I/tmp/pobj/gdal-3.11.0/gdal-3.11.0/port -I/tmp/pobj/gdal-3.11.0/build-amd64/port -I/tmp/pobj/gdal-3.11.0/gdal-3.11.0/frmts/zlib/contrib/infback9 -I/usr/local/include/json-c -I/usr/local/include -I/usr/local/include/libxml2 -O2 -pipe -fvisibility=hidden -DNDEBUG -std=gnu++17 -fPIC -fno-finite-math-only -Wall -Wextra -Winit-self -Wunused-parameter -Wmissing-declarations -Wshorten-64-to-32 -Wshadow -Wshadow-field -Wmissing-include-dirs -Wformat -Werror=format-security -Wno-format-nonliteral -Werror=vla -Wdate-time -Wnull-dereference -Wextra-semi -Wcomma -Wfloat-conversion -Wdocumentation -Wno-documentation-deprecated-sync -Wunused-private-field -Wnon-virtual-dtor -Woverloaded-virtual -Wsuggest-override -Wstring-conversion -fno-operator-names -Wzero-as-null-pointer-constant -Wold-style-cast -Weffc++ -MD -MT port/CMakeFiles/cpl.dir/cpl_conv.cpp.o -MF port/CMakeFiles/cpl.dir/cpl_conv.cpp.o.d -o port/CMakeFiles/cpl.dir/cpl_conv.cpp.o -c /tmp/pobj/gdal-3.11.0/gdal-3.11.0/port/cpl_conv.cpp
In file included from /tmp/pobj/gdal-3.11.0/gdal-3.11.0/port/cpl_conv.cpp:57:
In file included from /usr/include/c++/v1/set:1495:
In file included from /usr/include/c++/v1/functional:552:
In file included from /usr/include/c++/v1/__functional/boyer_moore_searcher.h:27:
In file included from /usr/include/c++/v1/vector:3023:
In file included from /usr/include/c++/v1/locale:232:
/usr/include/c++/v1/__locale_dir/locale_base_api/bsd_locale_fallbacks.h:109:15: error: use of undeclared identifier 'vasprintf'; did you mean 'vsprintf'?
  109 |   int __res = vasprintf(__s, __format, __va);
      |               ^
/usr/include/c++/v1/cstdio:133:9: note: 'vsprintf' declared here
  133 | using ::vsprintf _LIBCPP_USING_IF_EXISTS;
      |         ^
In file included from /tmp/pobj/gdal-3.11.0/gdal-3.11.0/port/cpl_conv.cpp:57:
In file included from /usr/include/c++/v1/set:1495:
In file included from /usr/include/c++/v1/functional:552:
In file included from /usr/include/c++/v1/__functional/boyer_moore_searcher.h:27:
In file included from /usr/include/c++/v1/vector:3023:
In file included from /usr/include/c++/v1/locale:232:
/usr/include/c++/v1/__locale_dir/locale_base_api/bsd_locale_fallbacks.h:109:25: error: cannot initialize a parameter of type 'char *' with an lvalue of type 'char **'
  109 |   int __res = vasprintf(__s, __format, __va);
      |                         ^~~
/usr/include/stdio.h:269:32: note: passing argument to parameter here
  269 | int      vsprintf(char * __restrict, const char * __restrict, __va_list);
      |                                    ^
2 errors generated.
ninja: build stopped: subcommand failed.
```

--- a/port/cpl_conv.cpp
+++ b/port/cpl_conv.cpp
@@ -15,8 +15,10 @@
 
 #ifdef HAVE_USELOCALE
 // For uselocale, define _XOPEN_SOURCE = 700
-// but on Solaris, we don't have uselocale and we cannot have
-// std=c++11 with _XOPEN_SOURCE != 600
+// and OpenBSD with libcxx 19.1.7 requires 800 for vasprintf
+// (cf https://github.com/OSGeo/gdal/issues/12619)
+// (not sure if the following is still up to date...) but on Solaris, we don't
+// have uselocale and we cannot have std=c++11 with _XOPEN_SOURCE != 600
 #if defined(__sun__) && __cplusplus >= 201103L
 #if _XOPEN_SOURCE != 600
 #ifdef _XOPEN_SOURCE
@@ -28,7 +30,7 @@
 #ifdef _XOPEN_SOURCE
 #undef _XOPEN_SOURCE
 #endif
-#define _XOPEN_SOURCE 700
+#define _XOPEN_SOURCE 800
 #endif
 #endif
 

